lib/internal/modules/esm/hooks.js

```diff
@@ -127,31 +131,42 @@ class Hooks {
    * Import and register custom/user-defined module loader hook(s).
    * @param {string} urlOrSpecifier
    * @param {string} parentURL
+   * @param {any} [data] Arbitrary data to be passed from the custom
+   * loader (user-land) to the worker.
    */
-  async register(urlOrSpecifier, parentURL) {
+  async register(urlOrSpecifier, parentURL, data) {
     const moduleLoader = require('internal/process/esm_loader').esmLoader;
     const keyedExports = await moduleLoader.import(
       urlOrSpecifier,
       parentURL,
       kEmptyObject,
     );
-    this.addCustomLoader(urlOrSpecifier, keyedExports);
+    return this.addCustomLoader(urlOrSpecifier, keyedExports, data);
   }
 
   /**
    * Collect custom/user-defined module loader hook(s).
    * After all hooks have been collected, the global preload hook(s) must be initialized.
    * @param {string} url Custom loader specifier
    * @param {Record<string, unknown>} exports
+   * @param {any} [data] Arbitrary data to be passed from the custom loader (user-land)
+   * to the worker.
    */
-  addCustomLoader(url, exports) {
+  addCustomLoader(url, exports, data) {
     const {
       globalPreload,
       resolve,
       load,
+      initialize,
     } = pluckHooks(exports);
 
     if (globalPreload) {
```

Comment:
We should ensure that the warning won’t be printed if the loader contains both `initialize` and `globalPreload`. Actually we should completely ignore `globalPreload` if `initialize` is provided. 
```suggestion
    if (globalPreload && initialize == null) {
```

Reply:
So the behaviour if they provide both is to ignore `globalPreload` entirely? Should we not at least warn that they're providing both and only one will be used?

Reply:
The assumption is that if they know about `initialize`, they know it's the replacement for `globalPreload` and they expect only one will be used. We have to do that otherwise there won't be a way to write a loader that works in both versions.